### PR TITLE
Fix GUI Image caching to consider images that have been created but n…

### PR DESCRIPTION
**Pull request by carolhmj**
…ot fully loaded.

Related forum issue:  https://forum.babylonjs.com/t/gui-editor-bug-parse-from-snippet-with-external-image-from-dropbox-svg/37411/

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/13440/merge

Test environment: 
https://babylonsnapshots.z22.web.core.windows.net/refs/pull/13440/merge/index.html

To test a playground add it to the URL, for example:

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/13440/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/13440/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/13440/merge
https://gui.babylonjs.com/?snapshot=refs/pull/13440/merge
https://nme.babylonjs.com/?snapshot=refs/pull/13440/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/13440/merge#BCU1XR#0

**Comment by bjsplat** [comment]
Visualization tests for webgl2 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/13440/merge/testResults/webgl2/index.html

If tests were successful afterwards, this report might not be available anymore.

File: packages/dev/gui/src/2D/controls/image.ts [hunk 1]
```diff
@@ -555,15 +558,30 @@ export class Image extends Control {
...
 
         this._domImage.onload = () => {
+            if (value) {
+                const cachedData = Image.SourceImgCache.get(value);
+                if (cachedData) {
+                    cachedData.loaded = true;
+                    for (const waitingImage of cachedData.waitingForLoad) {
+                        waitingImage._onImageLoaded();
+                    }
+                    cachedData.waitingForLoad = [];
```
Comment (RaananW) [inline packages/dev/gui/src/2D/controls/image.ts:581]:
```suggestion
                    cachedData.waitingForLoad.length = 0;
```

File: packages/dev/gui/src/2D/controls/image.ts [hunk 2]
```diff
@@ -555,15 +558,30 @@ export class Image extends Control {
...
+                this._onImageLoaded();
+            } else {
+                cachedData.waitingForLoad.push(this);
+            }
             return;
         }
         this._domImage = engine.createCanvasImage();
         if (value) {
-            Image.SourceImgCache.set(value, { img: this._domImage, timesUsed: 1 });
+            Image.SourceImgCache.set(value, { img: this._domImage, timesUsed: 1, loaded: false, waitingForLoad: [this] });
```
Comment (RaananW) [inline packages/dev/gui/src/2D/controls/image.ts:570]: why is the array needed?

**Review by RaananW — COMMENTED** [review]
Not sure I follow the code.

Is there a need for the array? why not just keep a reference of the image?

File: packages/dev/gui/src/2D/controls/image.ts (hunk 2 above)
Comment (carolhmj) [reply packages/dev/gui/src/2D/controls/image.ts:570]: There might be multiple images that access _domImage while it is trying to load, so they should have their own _onImageLoaded methods when the onload call resolves.

Comment (RaananW) [reply packages/dev/gui/src/2D/controls/image.ts:570]: Got it! makes sense to keep the reference to all functions, but is there a need for the entire class to be passed? Why not use an observable that is being notified when the dom image was loaded? In this case using the `addOnce` would make sense to make sure the observable doesn't keep unneeded references.

Comment (carolhmj) [reply packages/dev/gui/src/2D/controls/image.ts:570]: Oooh, that makes sense and saves space on the cache, will do it

Comment (carolhmj) [reply packages/dev/gui/src/2D/controls/image.ts]: @RaananW done!

**Review by RaananW — APPROVED** [review]

File: packages/dev/gui/src/2D/controls/image.ts [hunk 3]
```diff
@@ -53,7 +53,10 @@ export class Image extends Control {
         key: string;
     } = { data: null, key: "" };
 
-    public static SourceImgCache = new Map<string, { img: IImage; timesUsed: number }>();
+    /**
+     * Cache of images to avoid loading the same image multiple times
+     */
+    public static SourceImgCache = new Map<string, { img: IImage; timesUsed: number; loaded: boolean; waitingForLoadCallback: Array<() => void> }>();
```
Comment (RaananW) [inline packages/dev/gui/src/2D/controls/image.ts:59]: Can you add a visualization test for that? to make sure it will always work as expected?

**Comment by bjsplat** [comment]
Visualization tests for webgl1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/13440/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

File: packages/dev/gui/src/2D/controls/image.ts (hunk 3 above)
Comment (RaananW) [reply packages/dev/gui/src/2D/controls/image.ts:59]: Thanks a lot! :-)

**Pull request merged**